### DISCUSSION TITLE
LED candle 'on' command

### DISCUSSION THREAD

**Discussion by You69Man** [Arduino-IRremote/Arduino-IRremote#1081] [Help] [answered]
Hi all,

I'm trying to remote control a 'real wax LED candle with 10-Key remote control' (if you google this phrase, you'll find pictures of these candles and the control with one red 'on', one green 'off' and eight other black control keys).

Using the ReceiveDump example I've managed to get the protocol, addresses and commands of all ten keys, and also to reproduce most of the commands (dim, brighten, candle mode, light mode, off ...) with an IR transmitter LED and to actually control the candle.

However while most of the commands are working correctly the 'on' command (to switch on the candle)  doesn't. 
So, I can only switch on the candle with the original remote control, but not with the arduino library.

Some examples of the output of the ReceiveDump:

'Light' key:  (working)
Protocol=NEC Address=0xB708 Command=0xE Raw-Data=0xF10EB708 32 bits LSB first

'Candle' key:  (working)
Protocol=NEC Address=0xB708 Command=0xC Raw-Data=0xF30CB708 32 bits LSB first

'Off' key: (working)
Protocol=NEC Address=0xB708 Command=0x2 Raw-Data=0xFD02B708 32 bits LSB first

'On' key: (not working!)
Protocol=NEC Address=0xB708 Command=0x0 Raw-Data=0xFF00B708 32 bits LSB first

I actually don't see much difference between these commands and I'm treating them all equal when sending, like

```
uint16_t sAddress = 0xB708; 
uint16_t sRepeats = 5;
enum Command { cOn           = 0x00,
                             cOff           = 0x02,
                             cTimer2h    = 0x04,
                             cTimer4h    = 0x06,
                             cTimer6h    = 0x08,
                             cTimer8h    = 0x0A,
                             cModeCandle = 0x0C,
                             cModeLight  = 0x0E,
                             cDimLower   = 0x10,
                             cDimHigher  = 0x12
                          };
              
IrSender.sendNEC(sAddress, cModeLight, sRepeats);    // working
IrSender.sendNEC(sAddress, cModeCandle, sRepeats);  // working
IrSender.sendNEC(sAddress, cOff, sRepeats);  // working
IrSender.sendNEC(sAddress, cOn, sRepeats);  // not working!
```

Used hardware:
- Arduino Nano
- IR receiver module
- IR transmitter module (KY-005 with 220ohm)

Does anyone have an idea what the reason might be, or what I could try?
Thanks for any hint!

PS: What I'm also wondering is the changing of the second nibble in the raw data (right from the 0xF). Can it be that this is also part of the command? I've also tried to send 16 bit commands (0xFF00 for 'off') but unfortinately without success.

**Comment by ArminJo** [comment, marked as answer]
What is the output of ReceiveDemo for a quick and a long press (including repeats) of your remote?
**There are two NEC protocols in the wild**, which differ only in repeat handling/sending.
You can test the variant by using` IrSender.sendNEC2(...`.
Is there a minimal number of repeats, especially for the on command?

**Reply by You69Man** [reply]
@ArminJo
output of ReceiveDemo for short and long press...

"off" command:
Protocol=NEC Address=0xB708 Command=0x2 Raw-Data=0xFD02B708 32 bits LSB first
Protocol=NEC Address=0xB708 Command=0x2 Raw-Data=0xFD02B708 32 bits LSB first
Protocol=NEC Address=0xB708 Command=0x2 Repeat gap=41250us
Protocol=NEC Address=0xB708 Command=0x2 Repeat gap=97550us
Protocol=NEC Address=0xB708 Command=0x2 Repeat gap=96050us
...

"on" command:
Protocol=NEC Address=0xB708 Command=0x0 Raw-Data=0xFF00B708 32 bits LSB first
Protocol=NEC Address=0xB708 Command=0x0 Raw-Data=0xFF00B708 32 bits LSB first
Protocol=NEC Address=0xB708 Command=0x0 Repeat gap=41850us
Protocol=NEC Address=0xB708 Command=0x0 Repeat gap=96100us
Protocol=NEC Address=0xB708 Command=0x0 Repeat gap=97600us
...

This still tells me that it is a NEC protocol. So, I wasn't aware that there is also a NEC2 variant.

But I've tried as you said and just changed the commands to IrSender.sendNEC2()

And viola! That makes the difference!

All commands that have already worked still work, but now I have to reduce the repeats to 0 otherwise the candle recognizes them as multiple commands (before only one command was recognized despite the repetition).
And now also the 'on' command works :-) ! 
For the 'on' however I have to apply 1 (or 2?) repeats, and then the candle actually turns on! Apparently it needs some repetitions to wake up from some kind of sleep mode...!?

So, I'm very happy with that! Many thanks for the quick support!
Manfred

**Reply by ArminJo** [reply]
Thanks for the feedback and glad to be able to help you.
One last question, just to understand.
For a long press to "on", are you receiving one or two lines of
`Protocol=NEC Address=0xB708 Command=0x0 Raw-Data=0xFF00B708 32 bits LSB first`
before the Repeat's?

**Reply by You69Man** [reply]
For a long press of "on" as well as for a long press of "off" there is always exactly only one "Protocol=NEC..Raw-Data...." line", after which immediatly the "...Repeat..." lines appear. There is no difference besides the Command byte (and the second nibble in the Raw-Data, which meaning I still don't know)

**Reply by You69Man** [reply]
Maybe you are also interested in this timing dump:

"On"
Raw result in microseconds - with leading gap
rawData[68]: 
 -3276750
 +9150,-4450
 + 600,- 550 + 550,- 550 + 600,- 550 + 600,-1650
 + 600,- 550 + 550,- 550 + 600,- 550 + 550,- 550
 + 600,-1650 + 600,-1650 + 600,-1650 + 600,- 550
 + 600,-1650 + 600,-1650 + 550,- 550 + 600,-1650
 + 600,- 550 + 600,- 550 + 550,- 550 + 600,- 550
 + 600,- 500 + 600,- 550 + 600,- 550 + 550,- 550
 + 600,-1650 + 600,-1650 + 600,-1650 + 600,-1650
 + 600,-1650 + 600,-1650 + 600,-1650 + 600,-1650
 + 600
Sum: 67150

"Off"
Raw result in microseconds - with leading gap
rawData[68]: 
 -3276750
 +9200,-4400
 + 600,- 550 + 600,- 550 + 550,- 550 + 600,-1650
 + 600,- 550 + 550,- 600 + 550,- 550 + 600,- 550
 + 550,-1700 + 550,-1700 + 550,-1700 + 600,- 500
 + 600,-1650 + 600,-1650 + 600,- 550 + 600,-1650
 + 600,- 500 + 600,-1650 + 600,- 550 + 600,- 550
 + 600,- 500 + 600,- 550 + 600,- 550 + 550,- 550
 + 600,-1650 + 600,- 550 + 600,-1650 + 550,-1700
 + 600,-1650 + 550,-1700 + 550,-1700 + 550,-1650
 + 600
Sum: 67150

For me, I also don't see any significant difference which could explain the different behaviour between on and off.
But now that it works with NEC2, so no further investigation required from my side.

**Reply by ArminJo** [reply]
So it seems that your sender signal is to weak to "wake up" the integrated IR receiver and therefore the first protocol frame containing the key information is not completly received. This is fixed by using the NEC2 protocol which sends the first frame again as repeat,
Another solution is to [increase sending power](https://github.com/Arduino-IRremote/Arduino-IRremote#increase-strength-of-sent-output-signal) or **send any dummy signal first to "wake up" the integrated IR receiver** and then send the command with NEC protocol.

**Reply by You69Man** [reply]
This also came already into my mind, but now that it works with NEC2 I haven't tried anymore. 
But yes, I will try to reduce the 220ohm resistance next days and let you know if it has any effect with the standard NEC protocol!

**Reply by You69Man** [reply]
Ok, I've made some new tests now, and it is a bit strange.

First of all I have removed the serial resistor of the IR-LED (0 Ohm; I know about the 'risk' :) ). The used IR-LED is a white type (unfortunately I don't have the exact type or a datasheet, but the package says 1.35V, 100mA).
Removing the resistor doesn't make any improvement with NEC protocol, but the NEC2 still works perfectly. 

Then I used another IR-LED ("light-blue" type, also 1.35V, 100mA), also with and without serial resistor. 
This time (still independently of with or without serial resistor) the ON command works in approximately 3 of 4 cases with NEC protocol. And there is a depencency for a successfull ON command, with the durance of the OFF state. So, if the OFF state is only 100ms, than the probability of a successful ON is higher, if the OFF state is >1s, it shrinks dramatically.
Yet, with NEC2 and the blue IR-LED it always works perfectly.

Resumee: I can confirm that there is some kind of hardware dependency for a successful "wake up" (ON) command, at least using the NEC protocol. The type of IR LED has more influence that the serial resistor.
NEC2 however somehow always gets over this issue.

Best regards,
Manfred

**Reply by ArminJo** [reply]
Hi Manfred, 
thanks for your research!
Your results sounds real mysterious and does not give the clear result I hoped for.
So it is fine, that at least you have one working variant with NEC2 😀 .
Best regards
Armin
